### PR TITLE
[WIP] Refactor _fetch_file(s)

### PR DESCRIPTION
**Pull request by kchawla-pi**
While working on #1792 I found it painful & slow to read the _fetch_files() & _fetch_file_().
Additionally, the new URLs on OSF may necessitate changing how the passed in URLs are handled. 

As per issue #1769, I have refactored the function into several smaller ones, which have made it considerably easier for me to comprehend it, and modify it if necessary.

The logic has not been altered to maintain Python2 compatibility & make the refactoring quick, safe, and easy.
This is in line with industry standard clean code principles for improving code quality and something I intend to do simultaneously where ever I see code that is difficult to comprehend.

### CODE REVIEW

File: nilearn/datasets/utils.py [hunk 1]
```diff
@@ -774,6 +765,49 @@ def _fetch_files(data_dir, files, resume=True, mock=False, verbose=1):
     return files_
 
 
+def _download_or_mock_file(abort, dl_file, file_, mock, target_file,
+                           temp_target_file, url):
+    if mock:
```
Comment (bthirion) [inline nilearn/datasets/utils.py]: Could you add a one-line function description to explain in a nutshell what the function does ?

File: nilearn/datasets/utils.py [hunk 2]
```diff
@@ -774,6 +765,49 @@ def _fetch_files(data_dir, files, resume=True, mock=False, verbose=1):
...
+        if not mock or os.path.getsize(dl_file) != 0:
+            _uncompress_file(dl_file, verbose=verbose)
+        else:
+            os.remove(dl_file)
+    except Exception as e:
+        abort = str(e)
+    return abort
+
+
+def _move_dataset_files(dl_file, opts, temp_dir):
```
Comment (bthirion) [inline nilearn/datasets/utils.py]: (same for other private functions)

File: nilearn/datasets/utils.py [hunk 3]
```diff
@@ -774,6 +766,85 @@ def _fetch_files(data_dir, files, resume=True, mock=False, verbose=1):
...
+    return abort
+
+
+def _check_files_existence(file_1, file_2):
+    """ Returns True if both the provided files exist.
+    """
+    target_file_does_not_exist = not os.path.exists(file_1)
+    temp_target_file_does_not_exist = not os.path.exists(file_2)
+    file_does_not_exist = target_file_does_not_exist and temp_target_file_does_not_exist
+    return file_does_not_exist
```
Comment (GaelVaroquaux) [inline nilearn/datasets/utils.py]: 3-liner functions are in general bad practice. The reason being that when doing a careful read of the codebase, one needs to read them to know what they actually do (the name is not a contract, and bugs hide this).

I am not very enthousiastic about merging changes that turn 3 lines of code used twice into a function. It makes the code harder to read and maintain.

File: nilearn/datasets/utils.py [hunk 4]
```diff
@@ -774,6 +766,85 @@ def _fetch_files(data_dir, files, resume=True, mock=False, verbose=1):
...
+    None or String(exception)
+    """
+    try:
+        if not mock or os.path.getsize(file_) != 0:
+            _uncompress_file(file_, verbose=verbose)
+        else:
+            os.remove(file_)
+    except Exception as e:
+        abort = str(e)
+    return abort
```
Comment (GaelVaroquaux) [inline nilearn/datasets/utils.py]: This function is used only once, and understanding what it does is as difficult from the code as from the function name and description. It should not be factored out.

File: nilearn/datasets/utils.py [hunk 5]
```diff
@@ -774,6 +766,85 @@ def _fetch_files(data_dir, files, resume=True, mock=False, verbose=1):
...
+        path to the destination root directory
+        
+    dst_sub_dir: string
+        Name or path for the subdirectory within the destination root directory
+        
+    Returns
+    -------
+    dst_filepath: string
+        Path to the moved file.
+    """
```
Comment (GaelVaroquaux) [inline nilearn/datasets/utils.py]: Another function that is very simple and used once.

**Comment by GaelVaroquaux** [comment]
@kchawla-pi : I am really sorry, but these changes do not make the codebase easier to maintain, rather harder. I've worked on many codebases, and codebases that are structured like this (with many tiny functions used once or twice) are very hard to follow: they require jumping from one function to another all the time, and keeping in one's working memory what the functions do.

I am not in favor of merging these changes. Let's try to discuss this IRL.
